src/main/kotlin/org/rust/ide/annotator/RsCargoCheckAnnotator.kt

```diff
@@ -0,0 +1,203 @@
+import org.rust.cargo.toolchain.CargoTopMessage
+
+data class CargoCheckAnnotationInfo(val file: PsiFile, val editor: Editor)
+
+class CargoCheckAnnotationResult(commandOutput: List<String>, val project: Project): ModificationTracker {
+
+    private val modificationTracker = PsiManager.getInstance(project).modificationTracker
+    private val parser = JsonParser()
+    //private val LOG = Logger.getInstance(RsCargoCheckAnnotator::class.java)
+
+    val messages: List<CargoTopMessage> =
+        commandOutput.filter {
+            it.trimStart().startsWith("{")
+        }.filter {
+            val first = it.indexOfFirst { it == '"' }
+            assert(first != -1)
+            val second = it.indexOf('"', first + 1)
+            assert(second != -1)
+
+            when (it.subSequence(first + 1, second)) {
+                "message" -> true
+                "features" -> false
+                else -> {
+                    assert(false)
+                    throw AssertionError()
+                }
+            }
+        }.map {
+            parser.parse(it)
+        }.mapNotNull {
+            if (it.isJsonObject)
+                CargoTopMessage.fromJson(it.asJsonObject)
+            else
+                null
+        }
+
+    override fun getModificationCount(): Long {
+        val modificationCount = modificationTracker.modificationCount
+        //LOG.info("getModificationCount = $modificationCount")
+        return modificationCount
+    }
+}
+
+class RsCargoCheckAnnotator : ExternalAnnotator<CargoCheckAnnotationInfo, CargoCheckAnnotationResult>() {
+
+    private fun getCachedResult(file: PsiFile) =
+        CachedValuesManager.getManager(file.project).createCachedValue {
+            CachedValueProvider.Result.create(
+                checkProject(file),
+                PsiModificationTracker.MODIFICATION_COUNT)
+
+        }
+
+    override fun collectInformation(file: PsiFile, editor: Editor, hasErrors: Boolean): CargoCheckAnnotationInfo? {
+        return CargoCheckAnnotationInfo(file, editor)
+    }
+
+    override fun doAnnotate(info: CargoCheckAnnotationInfo) = getCachedResult(info.file).value
+
+    override fun apply(file: PsiFile, annotationResult: CargoCheckAnnotationResult?, holder: AnnotationHolder) {
+        annotationResult ?: return
+        val doc = holder.currentAnnotationSession.file.viewProvider.document ?: throw AssertionError()
+
+        for (topMessage in annotationResult.messages) {
+            val message = topMessage.message
+            val filePath = holder.currentAnnotationSession.file.virtualFile.path
+
+            val severity = when (message.level) {
+                "error" -> HighlightSeverity.ERROR
+                "warning" -> HighlightSeverity.WEAK_WARNING
+                else -> throw AssertionError()
+            }
+
+            val codeStr = message.code?.code
+
+            // If spans are empty we add a "global" error
+            if (message.spans.isEmpty()) {
```

Comment:
Though I see that the current implementation works fine with all the corner cases that I was able to remember. It seems that the compiler output is more friendly now, so I think this additional `span.extension` walking can be skipped, at least at this moment.